src/main/scala/pbdirect/PBWriter.scala

```diff
@@ -1,121 +1,223 @@
 package pbdirect
 
-import java.io.ByteArrayOutputStream
-
-import cats.{ Contravariant, Functor }
-import com.google.protobuf.CodedOutputStream
+import cats.{Contravariant, Functor}
+import com.google.protobuf.{CodedOutputStream, WireFormat}
+import pbdirect.LowPriorityPBWriterImplicits.SizeMap
 import shapeless.{:+:, ::, CNil, Coproduct, Generic, HList, HNil, Inl, Inr, Lazy}
 
 trait PBWriter[A] {
-  def writeTo(index: Int, value: A, out: CodedOutputStream): Unit
+  def writeTo(index: Int, value: A, out: CodedOutputStream, sizes: SizeMap): Unit
+  def writtenBytesSize(index: Int, value: A, sizes: SizeMap): Int
+}
+
+object LowPriorityPBWriterImplicits {
+  type SizeMap = java.util.Map[Any, Int]
+  
+  private[pbdirect] def memoize[A](writtenBytesSize: (Int, A, SizeMap) => Int): (Int, A, SizeMap) => Int = {
+    (index: Int, value: A, sizes: SizeMap) => {
+      if (sizes.containsKey(value)) {
+        sizes.get(value)
+      } else {
+        val size = writtenBytesSize(index, value, sizes)
+        sizes.put(value, size)
```

Comment:
I'm a bit worried here as the size depends on the index but the cached size is only indexed by the value. That's an issue in case of messages containing 2 identical values with different index.

Reply:
That's a great point.  Will fix and add a test case.  Hopefully the fix (which I think will involve changing the `sizes` map to a `Map[AnyRef, Map[Int, Int]]`) doesn't impact performance too much.